### PR TITLE
set-up codegen from sigstore bundle protobufs

### PR DESCRIPTION
**Pull request by bdehamer**
Signed-off-by: Brian DeHamer <bdehamer@github.com>

#### Summary

Adds some plumbing to automatically codegen types for the sigstore bundle format from the published protobufs.

Types can be re-generated by executing the following:
```
npm run codegen:bundle
```

The generated types are written to `src/bundle/__generated__`.

#### Release Note

* New script in `package.json` for re-generating bundle types from their protobufs

### CODE REVIEW

**Comment by bdehamer** [comment]
cc @kommendorkapten @patflynn generating TypeScript types from the protobufs in https://github.com/sigstore/cosign/pull/2204

File: bin/generate-bundle-types [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+curl -L --silent --show-error \
+  -o /tmp/defs/envelope.proto \
+  https://raw.githubusercontent.com/${REPO}/${REF}/${PB_PATH}/envelope.proto
+
+curl -L --silent --show-error \
+  -o /tmp/defs/bundle.proto \
+  https://raw.githubusercontent.com/${REPO}/${REF}/${PB_PATH}/bundle.proto
+
+protoc --plugin=./node_modules/.bin/protoc-gen-ts_proto \
```
Comment (feelepxyz) [inline bin/generate-bundle-types:18]: Does this need to be installed separately? Should we set up a Dockerfile for this?

Comment (feelepxyz) [reply bin/generate-bundle-types:18]: Realised running this in docker is not entirely trivial.. can have a crack at this if helpful as I kind of started poking at this while watching MS training videos 😄  would this be helpful?

Comment (bdehamer) [reply bin/generate-bundle-types:18]: @feelepxyz Yeah, if you wanna take a crack at this, that would be great. I wasn't sure that if asking the dev to have docker installed was any less burdensome than just installing `protoc` 😆

Comment (feelepxyz) [reply bin/generate-bundle-types:18]: > I wasn't sure that if asking the dev to have docker installed was any less burdensome than just installing `protoc` 

This is a good point, I think docker is probably fairly common though. I noticed that using an older version of protoc generated slightly different output which would not be the case if we pin the version in the docker image.

Will also add some docs/readme on installing docker as a prerequisite for generating the bundle protobufs.

File: bin/generate-bundle-types [hunk 2]
```diff
@@ -0,0 +1,29 @@
+#!/usr/bin/env sh
```
Comment (feelepxyz) [inline bin/generate-bundle-types:1]: I'm not 100% sold on the pattern here as it feels a bit surprising but went with something we did on a couple Dependabot projects. Here all you need to do is have docker installed and run `bin/generate-bundle-types` and should "just work".

I moved the curl downloads out from and set them up in the `Dockerfile` so we can run this bin script without network access 🔒

File: Dockerfile.bundle-types [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  && rm /tmp/protoc.zip
+
+###############
+FROM node:16.17
+
+WORKDIR /client
+
+# Download proto definitions for .sigstore bundle from sigstore/cosign
+ARG REPO=sigstore/cosign
+ARG REF=a8ea54021bb6f850b30ee22bd15e0477d623e00d
```
Comment (feelepxyz) [inline Dockerfile.bundle-types:24]: Bleeding edge 😄  looks like this is referencing https://github.com/sigstore/cosign/pull/2204/files

File: README.md [hunk 4]
```diff
@@ -22,6 +22,33 @@ A tool for signing and verifying npm packages.
 * Support offline and online signature verification
 * Retrieve root cert and public keys via TUF client
 
+## Prerequisites
+
+- Node.js version 16+ (LTS)
+- Add `node_modules/.bin` to your `$PATH` to access the installed `tsc` bin
```
Comment (feelepxyz) [inline README.md:28]: @bdehamer added these basic setup requirements while adding the docker prereq. Is this what you ended up doing as well with `PATH`?

**Comment by feelepxyz** [comment]
@bdehamer just force pushed as I forgot to add sig-off info on the commits.

**Comment by feelepxyz** [comment]
@eddiezane 👋 gentle bump, did you have any ideas for improvements here?

File: README.md (hunk 4 above)
Comment (bdehamer) [reply README.md:28]: I've never done this. I think the convention here is to just use `npx` (`npm exec`) to invoke project-local binaries.

File: hack/images/protoc/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+###############
+FROM node:16.17
+
+WORKDIR /client
+
+COPY --from=protoc /protoc/bin/protoc /usr/local/bin
+
+ARG TSPROTO_VERSION=1.125.0
+RUN npm install -g "ts-proto@${TSPROTO_VERSION}"
```
Comment (feelepxyz) [inline hack/images/protoc/Dockerfile:25]: Installing this explicitly is much nicer 👍

**Review by feelepxyz — APPROVED** [review]
@bdehamer ace refactor of my codes 👍 💯  much easier to follow.

**Pull request merged**